### PR TITLE
Change swagger processors to support data-services

### PR DESCRIPTION
**Pull request by GDLMadushanka**
Add functionality to create swagger from AxisResourceMap of a data
service
Change swagger JSON / YAML processors to support data-services
Fixes wso2/product-ei/issues/4999

### CODE REVIEW

File: components/mediation-commons/src/main/java/org/wso2/carbon/mediation/commons/rest/api/swagger/ServerConfig.java [hunk 1]
```diff
@@ -28,10 +28,19 @@
 public interface ServerConfig {
 
     /**
-     * Function to retrieve server host name for host entry in swagger definition
+     * Function to retrieve server host name for host entry in swagger definition (API)
```
Comment (sajinieKavindya) [inline components/mediation-commons/src/main/java/org/wso2/carbon/mediation/commons/rest/api/swagger/ServerConfig.java]: Shall we end this sentence with a period?

File: components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/CarbonServerConfig.java [hunk 2]
```diff
@@ -112,4 +112,8 @@ public String getHost(API api) throws AxisFault {
             throw new AxisFault("http/https transport listeners are required in axis2.xml");
         }
     }
+
+    @Override public String getHost(String transport) throws AxisFault {
+        return null;
```
Comment (sdkottegoda) [inline components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/CarbonServerConfig.java]: Shall we add a comment mentioning why we are returning null here?

Comment (GDLMadushanka) [reply components/mediation-admin/org.wso2.carbon.rest.api/src/main/java/org/wso2/carbon/rest/api/CarbonServerConfig.java]: +1 resolved

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerGenerator.java [hunk 3]
```diff
@@ -142,47 +142,13 @@ protected String retrieveFromRegistry(API api, CarbonHttpRequest request) throws
...
-
-        } else {
-            if (resourcePath.startsWith(CONFIG_REG_PREFIX)) {
-                resourcePath = resourcePath.substring(5);
-            } else {
-                resourcePath = resourcePath.substring(4);
-                isSourceConfigReg = false;
-            }
-        }
+            resourcePath = CONFIG_REG_PREFIX + resourcePathBuilder.toString();
```
Comment (milindaperera) [inline components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerGenerator.java]: You can refactor removing this "CONFIG_REG_PREFIX" constant and use one defined in SwaggerProcessorConstants.java. Also refactor usage of CONFIG_REG_PREFIX in SwaggerGenerator.java

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerJsonProcessor.java [hunk 4]
```diff
@@ -43,32 +44,45 @@
...
     public void process(CarbonHttpRequest request, CarbonHttpResponse response,
                         ConfigurationContext configurationContext) throws AxisFault {
-        API api = getAPIFromSynapseConfig(request);
 
-        if (api == null) {
+        API api = getAPIFromSynapseConfig(request);
+        String responseString = null;
+        if (api != null) {
+            responseString = processAPI(api, request);
+        } else if (request.getContextPath().contains("/" + SwaggerProcessorConstants.SERVICES_PREFIX)) {
```
Comment (milindaperera) [inline components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerJsonProcessor.java]: the format is 
host:port/services/blabla
host:port/t/tenantDomain/services/blabla
but this condition also pick
host:port/blabla/services

Does this get handled in SwaggerUtils.takeDataServiceSwagge

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/utils/SwaggerUtils.java [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    /**
+     * Fetch the swagger from registry if available or create one from scratch.
+     *
+     * @param requestURI           request URI.
+     * @param configurationContext Configuration context with details.
+     * @param isJSON               result format JSON or YAML.
+     * @return Swagger definition as string
+     * @throws AxisFault Error occurred while fetching the host details.
+     */
+    public static String takeDataServiceSwagger(String requestURI, ConfigurationContext configurationContext,
```
Comment (milindaperera) [inline components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/utils/SwaggerUtils.java]: let's rename takeDataServiceSwagger => getDataServiceSwagger

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/utils/SwaggerUtils.java [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.mediation.transport.handlers.requestprocessors.swagger.format.utils;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import io.swagger.models.*;
```
Comment (milindaperera) [inline components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/utils/SwaggerUtils.java]: We need to add all imports specifically

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/CarbonServerConfig.java [hunk 7]
```diff
@@ -60,57 +60,69 @@ public String getHost(API api) throws AxisFault {
...
+
+    private String getHostNameFromTransport(String transport) throws AxisFault {
+        //Retrieve WSDLPrefix to retrieve host
+        //If transport is limited to https, https host will generate. Otherwise http will be generated
+        TransportInDescription transportIn = axisConfiguration.getTransportIn(transport);
+
+        if (transportIn != null) {
+            // Give priority to WSDLEPRPrefix
+            if (transportIn.getParameter(SwaggerConstants.WSDL_EPR_PREFIX) != null) {
+                String wsdlPrefixParam = (String) transportIn.getParameter(SwaggerConstants.WSDL_EPR_PREFIX).getValue();
```
[the remark this answers is not in the document]
Comment (GDLMadushanka) [reply components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/CarbonServerConfig.java]: When WSDLPrefix is empty passthrough transport is not accepting traffic and it does not come to this point.

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerGenerator.java (hunk 3 above)
Comment (GDLMadushanka) [reply components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/SwaggerGenerator.java]: +1 fixed

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/utils/SwaggerUtils.java (hunk 5 above)
Comment (GDLMadushanka) [reply components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/utils/SwaggerUtils.java]: +1 fixed

File: components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/utils/SwaggerUtils.java (hunk 6 above)
Comment (GDLMadushanka) [reply components/mediation-initializer/org.wso2.carbon.mediation.transport.handlers/src/main/java/org/wso2/carbon/mediation/transport/handlers/requestprocessors/swagger/format/utils/SwaggerUtils.java]: +1 fixed

**Pull request merged**